lib/matplotlib/widgets.py

```diff
@@ -401,9 +408,18 @@ def observers(self):
 
     def _value_in_bounds(self, val):
         """Makes sure *val* is with given bounds."""
-        if self.valstep:
+        if isinstance(self.valstep, Number):
             val = (self.valmin
                    + round((val - self.valmin) / self.valstep) * self.valstep)
+        elif self.valstep is not None:
+            valstep = np.asanyarray(self.valstep)
+            if valstep.ndim != 1:
+                raise ValueError(
+                    f"valstep must have 1 dimension but has {valstep.ndim}"
+                )
+            idx = np.searchsorted(np.sort(valstep), val, side="left")
+            candidates = valstep[[idx-1, min(idx, valstep.size-1)]]
```

Comment:
I don't understand how this indexing works when you search using a sorted array, and then index the original unsorted one.

Reply:
Good catch thank you!

This was a remnant of the original implementation. This is far faster if you presort the array in the `init` method, but now that we're not doing that using `argmin` will be simpler and likely as fast.

Reply:
It still worked in the all the examples and tests because every example was sorted.